File: Engine/Android/core/src/main/kotlin/despia/engine/Source.kt
//
//  Source.kt — the JVM kernel provenance primitive (`dsx.source`).
//
//  This is the Kotlin twin of Engine/iOS/Source.swift.  Owners publish one
//  independently keyed plane into the app-wide reactive store:
//
//    source.<plane> = { state: never|stale|live,
//                       serving: origin|cache|bundle,
//                       at: <ISO-8601 instant> }
//
//  Persistence is deliberately injected.  The pure kernel has no Android Context
//  and no desktop filesystem policy; each process host installs its durable stamp
//  store.  The in-memory default remains total for pure JVM/tests and honestly loses
//  "ever loaded" state at process exit rather than inventing persistence.
//
//  The KERNEL FACTS — `source.boot` (first|warm) and `source.online` — ride `seed`/
//  `setOnline` below.  Source.swift seeds both from `DSXBoot` and keeps `online` live
//  off an `NWPathMonitor` it owns; this kernel tier compiles SDK-free, so reachability
//  is INJECTED the same way persistence is: the process host seeds an initial value and
//  pushes every change (`:platform` SourceBackend.kt is the Android installer).  No
//  per-plane slice is written here — those stay OWNER-published (`track`/`publish`).
//

package despia.engine

import java.security.MessageDigest
import java.time.Clock
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.concurrent.ConcurrentHashMap

/** Minimal persistence face for provenance's per-plane/per-origin first-load stamp. */
interface DSXSourceStampStore {
    fun read(key: String): String?
    fun write(key: String, value: String)
}

private class MemoryDSXSourceStampStore : DSXSourceStampStore {
    private val values = ConcurrentHashMap<String, String>()
    override fun read(key: String): String? = values[key]
    override fun write(key: String, value: String) { values[key] = value }
}

object DSXSource {
    const val never = "never"
    const val stale = "stale"
    const val live = "live"
    const val servingOrigin = "origin"
    const val servingCache = "cache"
    const val servingBundle = "bundle"

    private val lock = Any()
    private const val MAX_KEY_BYTES = 8 * 1_024
    @Volatile private var stamps: DSXSourceStampStore = MemoryDSXSourceStampStore()
    @Volatile internal var clock: Clock = Clock.systemUTC()

    /** The reserved stamp identity behind `source.boot` — the `dsx.source.booted` twin. */
    private const val BOOT_PLANE = "boot"
    private const val BOOT_KEY = "install"

    /** Install the host-owned durable store. Safe to call again during process boot. */
    fun installStampStore(store: DSXSourceStampStore?) {
        synchronized(lock) { stamps = store ?: MemoryDSXSourceStampStore() }
    }

    /**
     * Seed the KERNEL facts before any surface renders (Source.swift `seed()`): `source.boot`
     * = "first" until this install has booted once, "warm" forever after, and `source.online`
     * = the host's current reachability answer. Idempotent within a process; the boot stamp is
     * written on the FIRST call, so a second call in the same install already reads "warm".
     *
     * DIVERGENCE from Swift, by construction: Source.swift arms its own `NWPathMonitor` here.
     * This tier has no Android/JVM reachability API (it compiles SDK-free), so the host seeds
     * the value and pushes changes through `setOnline` — the same injection the stamp store
     * uses. Neither call writes a per-plane slice: those stay owner-published.
     *
     * `online = null` means the host CANNOT observe reachability (no permission, no service):
     * `source.online` is then left ABSENT rather than guessed, because a fact that lies is
     * worse than a fact that is missing (Article 7). `source.boot` still seeds.
     */
    fun seed(online: Boolean?) {
        val at = DateTimeFormatter.ISO_INSTANT.format(Instant.now(clock))
        val warm = synchronized(lock) {
            val had = readStampLocked(BOOT_PLANE, BOOT_KEY) != null
            if (!had) runCatching { stamps.write(stampKey(BOOT_PLANE, BOOT_KEY), at) }
            had
        }
        DSX.state.setPath("source.boot", if (warm) "warm" else "first")
        if (online != null) DSX.state.setPath("source.online", online)
    }

    /**
     * Reachability changed (the `pathUpdateHandler` twin). DEDUPED exactly like Swift: an
     * unchanged value neither re-publishes nor re-fires `source.changed`, so a flapping
     * transport can't spam every `visible-if` bound to the plane. Call on the main thread —
     * the installer hops for you.
     */
    fun setOnline(online: Boolean) {
        if ((DSX.state.getPath("source.online") as? Boolean) == online) return
        DSX.state.setPath("source.online", online)
        ModuleRegistry.shared.dispatch("source.changed", mapOf("plane" to "online", "state" to online), combine = ModuleRegistry.Combine.void)
    }

    /** Seed one owner plane before its first load this session. */
    fun track(plane: String, key: String) {
        val safePlane = validatedPlane(plane) ?: return
        val safeKey = validatedKey(key) ?: return
        val state = synchronized(lock) {
            if (readStampLocked(safePlane, safeKey) == null) never else stale
        }
        write(safePlane, mapOf("state" to state), fire = false)
    }

    /** Report bytes which are actually serving now. A fresh origin answer stamps the install. */
    fun publish(
        plane: String,
        serving: String,
        fresh: Boolean,
        key: String,
        meta: Map<String, Any?> = emptyMap(),
    ) {
        val safePlane = validatedPlane(plane) ?: return
        val safeKey = validatedKey(key) ?: return
        if (serving !in setOf(servingOrigin, servingCache, servingBundle)) return
        val at = DateTimeFormatter.ISO_INSTANT.format(Instant.now(clock))
        val state = synchronized(lock) {
            var stamp = readStampLocked(safePlane, safeKey)
            if (fresh && stamp == null) {
                runCatching { stamps.write(stampKey(safePlane, safeKey), at) }
                // Treat this successful origin serve as live even if durable storage is
                // unavailable. Persistence failure must not make current-screen truth lie.
                stamp = at
            }
            if (fresh) live else if (stamp == null) never else stale
        }
        val slice = LinkedHashMap<String, Any?>()
        slice["state"] = state
        slice["serving"] = serving
        slice["at"] = at
        meta.forEach { (name, value) ->
            if (name !in setOf("state", "serving", "at")) slice[name] = value
        }
        write(safePlane, slice, fire = true)
    }

    fun state(plane: String): String =
        ((DSX.state.getPath("source.$plane") as? Map<*, *>)?.get("state") as? String).orEmpty()

    private fun write(plane: String, slice: Map<String, Any?>, fire: Boolean) {
        DSX.state.setPath("source.$plane", slice)
        if (fire) ModuleRegistry.shared.dispatch(
            "source.changed",
            mapOf("plane" to plane, "state" to slice["state"]), combine = ModuleRegistry.Combine.void
        )
    }

    private fun validatedPlane(value: String): String? = value.trim().takeIf {
        it.matches(Regex("[A-Za-z][A-Za-z0-9_-]{0,63}"))
    }

    private fun validatedKey(value: String): String? = value.takeIf {
        it.toByteArray(Charsets.UTF_8).size <= MAX_KEY_BYTES
    }

    private fun readStampLocked(plane: String, key: String): String? =
        runCatching { stamps.read(stampKey(plane, key)) }.getOrNull()?.takeIf(String::isNotBlank)

    /** Fixed-size/non-secret persistence key; origins never become preference key material. */
    internal fun stampKey(plane: String, key: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest("$plane\u0000$key".toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
        return "v1.$digest"
    }

    internal fun resetForTests() {
        synchronized(lock) {
            stamps = MemoryDSXSourceStampStore()
            clock = Clock.systemUTC()
        }
    }
}

/** The owner-facing Kotlin facade, matching Swift's `dsx.source`. */
class DSXSourceFace {
    fun track(plane: String, key: String) = DSXSource.track(plane, key)
    fun publish(
        plane: String,
        serving: String,
        fresh: Boolean,
        key: String,
        meta: Map<String, Any?> = emptyMap(),
    ) = DSXSource.publish(plane, serving, fresh, key, meta)

    fun state(plane: String): String = DSXSource.state(plane)
    val online: Boolean get() = (DSX.state.getPath("source.online") as? Boolean) ?: true
}

File: Engine/iOS/Source.swift
//
//  Source.swift — the kernel PROVENANCE primitive (`dsx.source`).
//
//  ONE reactive answer per remote-loaded plane: has THIS INSTALL ever loaded it, what is on
//  screen right now, and when — so "installed online, first opened offline" is knowable state,
//  not a guess. The slice shape, per plane (`source.<plane>` in the ONE reactive store):
//
//      { state:   "never" | "stale" | "live",     // ever loaded (persisted) / fresh this session
//        serving: "origin" | "cache" | "bundle",  // what is actually on screen right now
//        at:      "2026-07-17T09:41:00Z" }        // last successful load/check (ISO-8601)
//
//  plus the kernel-owned facts `source.online` (reactive reachability) and `source.boot`
//  ("first" | "warm" — first launch of this install). Markup reads `dsx.source.web.state`
//  (JSE alias → `global.source.*`), the page watches `despia.global.watch("source", …)`
//  (a plain store slice — no new bridge), native hooks `source.changed`.
//
//  ARCHITECTURE (the ruling, argued in architecture/proposals/source-plane.md):
//  • `source.*` is KERNEL VOCABULARY — a reserved namespace like `nav`/`screen`/`dsx`
//    (Article 1: an alias for a state path; grants its maintainers nothing). This file is
//    MECHANISM ONLY: it names no module and no plane — "web"/"routes"/"content" are strings
//    the OWNING modules bring (Dom, Routing, the content plane), each publishing ONLY its
//    own slice through the `dsx.source` facade. No aggregator module exists — that would be
//    the shim Article 3 bans — and an excluded owner simply leaves its slice absent
//    (Article 7: degrade, never lie).
//  • CACHING stays ONE primitive (`dsx.content`); this plane is the thin REPORTING face it
//    (and the non-byte planes) feed. Anything that caches host bytes should ride
//    `dsx.content` and gets its provenance published for free (ContentStore feeds
//    `source.content` at its publish/confirmed-fresh seams).
//  • "Ever loaded" persists in `UserDefaults.standard` keyed per plane+origin — deliberately
//    NOT the App Group container (writes drop on unprovisioned builds) and NOT Caches (the
//    OS may purge it, which would regress `stale` → `never`). An environment switch changes
//    the key, so a fresh origin honestly reads `never` (the content plane's own per-origin
//    convention).
//
import Foundation
import Network

public enum DSXSource {

    /// The wire states (Article 8 — names, not platform types).
    public static let never = "never", stale = "stale", live = "live"
    public static let servingOrigin = "origin", servingCache = "cache", servingBundle = "bundle"

    private static let monitor = NWPathMonitor()
    private static var monitoring = false
    private static let iso = ISO8601DateFormatter()   // thread-safe per docs

    // MARK: - boot seed (kernel facts only — called once from DSXBoot)

    /// Publish `source.online` + `source.boot` and arm the reachability observer. Kernel facts
    /// only; every per-plane slice is owner-published via `track`/`publish` below.
    public static func seed() {
        let booted = UserDefaults.standard.bool(forKey: "dsx.source.booted")
        DSX.state.setPath("source.boot", booted ? "warm" : "first")
        if !booted { UserDefaults.standard.set(true, forKey: "dsx.source.booted") }
        DSX.state.setPath("source.online", InternetConnectionManager.isConnectedToNetwork())
        guard !monitoring else { return }
        monitoring = true
        monitor.pathUpdateHandler = { path in
            let online = path.status == .satisfied
            DispatchQueue.main.async {
                guard (DSX.state.getPath("source.online") as? Bool) != online else { return }
                DSX.state.setPath("source.online", online)
                ModuleRegistry.shared.dispatch("source.changed", ["plane": "online", "state": online], .void)
            }
        }
        monitor.start(queue: DispatchQueue(label: "dsx.source.reachability"))
    }

    // MARK: - owner verbs (reached through the `dsx.source` facade)

    /// Seed a plane BEFORE its first load of the session: `never` (this install has never
    /// loaded it under `key`) or `stale` (loaded before; nothing fresh yet). Call at the
    /// owner's launch so the state exists the moment markup first reads it.
    public static func track(_ plane: String, key: String) {
        let state = stamp(plane, key) == nil ? never : stale
        write(plane, ["state": state], fire: false)
    }

    /// A serve/load happened. `fresh: true` = the ORIGIN answered this session → `live` +
    /// the persisted first-load stamp (what makes `never` truthful forever). `fresh: false`
    /// = serving a cache/bundle copy → `stale`/`never` derived from the stamp, `serving` +
    /// `at` still updated (the "what's on screen" half stays honest offline). Extra `meta`
    /// keys (e.g. the content plane's `path`) merge into the slice.
    public static func publish(_ plane: String, serving: String, fresh: Bool,
                               key: String, meta: [String: Any] = [:]) {
        if fresh, stamp(plane, key) == nil {
            UserDefaults.standard.set(iso.string(from: Date()), forKey: stampKey(plane, key))
        }
        var slice: [String: Any] = ["state": fresh ? live : (stamp(plane, key) == nil ? never : stale),
                                    "serving": serving,
                                    "at": iso.string(from: Date())]
        meta.forEach { slice[$0.key] = $0.value }
        write(plane, slice, fire: true)
    }

    // MARK: - plumbing

    private static func write(_ plane: String, _ slice: [String: Any], fire: Bool) {
        let apply = {
            DSX.state.setPath("source.\(plane)", slice)
            if fire { ModuleRegistry.shared.dispatch("source.changed", ["plane": plane, "state": slice["state"] ?? ""], .void) }
        }
        Thread.isMainThread ? apply() : DispatchQueue.main.async(execute: apply)
    }

    private static func stamp(_ plane: String, _ key: String) -> String? {
        UserDefaults.standard.string(forKey: stampKey(plane, key))
    }
    private static func stampKey(_ plane: String, _ key: String) -> String {
        "dsx.source.\(plane).\(key)"
    }
}

/// The `dsx.source` face on the bus handle — how MODULES reach the primitive (the binding
/// rule: only `dsx`, never a kernel singleton). Publishing is owner-only by convention —
/// `source.*` is a reserved namespace; write someone else's plane and the state inspector
/// will out you. Reads are ordinary reactive store reads (`dsx.global.source…` native,
/// `dsx.source.*` in markup, `despia.global.watch("source", …)` on the page).
public struct DSXSourceFace {
    public func track(_ plane: String, key: String) { DSXSource.track(plane, key: key) }
    public func publish(_ plane: String, serving: String, fresh: Bool,
                        key: String, meta: [String: Any] = [:]) {
        DSXSource.publish(plane, serving: serving, fresh: fresh, key: key, meta: meta)
    }
    /// Typed read of one plane's state ("never"/"stale"/"live"; "" = plane absent).
    public func state(_ plane: String) -> String {
        ((DSX.state.getPath("source.\(plane)") as? [String: Any])?["state"] as? String) ?? ""
    }
    /// The kernel reachability fact (`source.online`).
    public var online: Bool { (DSX.state.getPath("source.online") as? Bool) ?? true }
}
